### ISSUE TITLE
Bug report with all "kommun" that the script can't render

### ISSUE THREAD

**Issue by matthiasfeist** [atorger/nvdb2osm#14]
Hi @atorger !
Thanks for your time and effort in producing this script! I have now processed all kommun-files from trafikverket (it just took 12 hours 😅) and there are quite some files that the script can not process.

I ran the script with the -d flag and here are all the log outputs. I hope there are come common problems so it's easy to fix :)
[debug-logs.zip](https://github.com/atorger/nvdb2osm/files/6074763/debug-logs.zip)

Let me know if I can help somehow!

And again thanks for the amazing work!

**Comment by matthiasfeist** [comment]
Haha, no problem! 
Just let me know if you want to get a new batch with maybe some fixes. I'm happy to re-run my AWS setup for you.

**Comment by atorger** [comment]
I've fixed some of the simpler bugs, now it's on to the NVDB geometry quirks... here's the geometry the script gets stuck on in Halmstad, the reference geometry is black, then a later layer positions it at the red, and note the disrupted highway. Hmm...

The easy way out is just to discard the items which are so poorly aligned that they don't snap. I'll experiment some and see where it lands.

![image](https://user-images.githubusercontent.com/13962807/109849028-334fe000-7c51-11eb-8ab7-d29e87972587.png)

**Comment by matthiasfeist** [comment]
Oh wow, that's a challenge.
I wonder if there is a feeback-channel to trafikverket somewhere. To tell them to fix their data :)

**Comment by matthiasfeist** [comment]
I agree maybe it makes more sense to leave these poorly aligned things out of the osm file.
Or alternatively, import both but make it very clear somehow that the user has to resolve this manually...

**Comment by atorger** [comment]
I have almost fixed all bugs (hopefully). The result of the quirky geometry above becomes as in this picture, ie a road becomes disconnected, but the JOSM Validator will warn the user about that so I think it's not too bad, especially as I have only seen this in one place so far.

I'm still working on the final fix though, I'll post again when it's a suitable time to make a new batch run and see how it goes.

![image](https://user-images.githubusercontent.com/13962807/110031506-4a68fd80-7d37-11eb-9e52-80934b6ccf43.png)

**Comment by atorger** [comment]
Heh, was another problem in the geometry, which happened to be just beside, see the disconnection. Fortunately this is the first time I see this type of error. I have noticed some minor offset errors in some places in NVDB geometry where one can see that there is a small bend in the middle of a road that should be straight, seems to be where segments are joined and I guess they may have used different measurement methods. However in all those cases the segments are still connected, it's only here I've seen this disconnect.

![image](https://user-images.githubusercontent.com/13962807/110032347-51dcd680-7d38-11eb-9263-48d1048a7755.png)

And this is how the same section looks on the website hitta.se, clearly they are using the same data :-). However, Lantmäteriets raster map does not have this artifact, and I have noted that they have more information and newer information than is in NVDB (ie if NVDB and the raster map differs, the raster map is nearly always right), but overall, NVDB is good.

![image](https://user-images.githubusercontent.com/13962807/110033370-961ca680-7d39-11eb-894c-6875cc65c7d3.png)

**Comment by atorger** [comment]
Now I have pushed the last fix, so now all the failed municipalities should work and you can try a new batch run.

I haven't tested them all though, but some are common errors. Sometimes when fixing bugs one reintroduce something other so it could happen that some municipality that worked before have stopped working. Hopefully not though...

Let me know how it goes.

Although failed runs are prioritized of course, it would also be nice if you could post a collected log output for all runs, also those that succeed. Analyzing the logs, especially the warning messages, for successful runs can be useful for me to find things that can be improved in the translation.

**Comment by matthiasfeist** [comment]
Interesting that lantmäteriet has bettersps than Trafikverket when it comes to roads. I really hope that they'll opensource their data soon as well.

I'm currently not at home but I'll start a new run on Sunday so I can post all logs on Monday 🙂

**Comment by atorger** [comment]
FYI, the script now parses railway crossings as well, and to be able to place them correctly (poor alignment in NVDB, and crossings need to be multiplied if there are several railway tracks) the shapefile with the national railway network needs to be provided. It's one file for whole Sweden so the same file is used for all muncipalities. Add the --railway_file argument, like this:

nvdb2osm.py -v --railway_file=Järnvägsnät_grundegenskaper.zip Luleå_kommun.zip luleå.osm

**Comment by atorger** [comment]
There is no parallelization in the script, ie only one CPU core will be used, so if you have lots of RAM you could convert several municipalities in parallel (probably no more than one per CPU core makes sense though) in your batch script to speed up processing. It will still take a bunch of hours though, so it's maybe not worth the extra complexity.

**Comment by matthiasfeist** [comment]
Hi @atorger. I've now run all the files with your latest version of the script and packed all the output for you here: https://nvdb-osm-map-data.s3.eu-north-1.amazonaws.com/Archive.zip (let me know when you have downloaded this file so I can delete it). 

I have my automatic pipeline ready now as well. It runs your script once per month with the latest downloads from Lastkajen. The output is then generated here: https://nvdb-osm-map-data.s3.eu-north-1.amazonaws.com/data-index.json
Note that not all "kommun" are available in Lastkajen for automatic download and it seems that I still have some bugs in my script. So sometimes the logs are missing even though there is an osm file... 

If you want to see the code of the pipeline project: https://github.com/matthiasfeist/nvdb-osm-diff-map

maybe that could also be interesting for @NKAmapper?

**Comment by atorger** [comment]
Thanks! I have downloaded the file so you can delete it.

**Comment by matthiasfeist** [comment]
great!

**Comment by atorger** [comment]
I've started to look at the failed runs from the latest batch run and fix bugs, I'll post an update here when I've fixed them so it's ready for a new run.

(Note that you don't need to run with debug activated, verbose is enough (-v) which should make the log files substantially smaller. When something fails I re-run it with debug anyway, and for those that succeed verbose is enough as I get the warning and info messages which is the only thing needed in terms of figuring out other things that may need improvement.)

**Comment by matthiasfeist** [comment]
OK, I'll change that :)

**Comment by atorger** [comment]
Now all the failed municipalities from the last run pass, and hopefully I haven't broken anything for the others... so you can make another batch run if you like.

**Comment by matthiasfeist** [comment]
I ran the pipeline again now.
The results are in this json file: https://nvdb-osm-map-data.s3.eu-north-1.amazonaws.com/data-index.json all the entries where osm is null but log is an object are instances where the script failed.

Here are the logs of the files that failed:
Bräcke kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/bracke-kommun.log
Eksjö kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/eksjo-kommun.log
Eskilstuna kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/eskilstuna-kommun.log
Fagersta kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/fagersta-kommun.log
Faluns kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/faluns-kommun.log
Hällefors kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/hallefors-kommun.log
Härjedalens kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/harjedalens-kommun.log
Laxå kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/laxa-kommun.log
Munkfors kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/munkfors-kommun.log
Nora kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/nora-kommun.log
Norrtälje kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/norrtalje-kommun.log
Partille kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/partille-kommun.log
Piteå kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/pitea-kommun.log
Ragunda kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/ragunda-kommun.log
Sundbybergs kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/sundbybergs-kommun.log
Surahammars kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/surahammars-kommun.log
Tanums kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/tanums-kommun.log
Tibro kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/tibro-kommun.log
Tidaholms kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/tidaholms-kommun.log
Timrå kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/timra-kommun.log
Tjörns kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/tjorns-kommun.log
Upplands-Bro kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/upplands-bro-kommun.log
Vimmerby kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/vimmerby-kommun.log
Älvkarleby kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/alvkarleby-kommun.log
Örnsköldsviks kommun https://nvdb-osm-map-data.s3.amazonaws.com/osm/ornskoldsviks-kommun.log

maybe it makes also sense to include a version number into the logs so that you know what version was actually run...

**Comment by atorger** [comment]
Yes I'll fix that.

I thought for a while all should pass this time, but no... seems to be the endpoint snapping stuff, when I fix it for some file it brakes for others. Let's see if I can succeed making a permanent fix this time...

**Comment by NKAmapper** [comment]
This is a great initiative. Will be very interesting to see the files. Big thanks to both of you.

With this full scale initiative, we are going to need an import plan and to get it through the review process, to "legalise" anyone's import. It would also be good to get the tagging questions resolved. I have taken the liberty of writing a draft plan just to get started. It needs further editing to be completed, in particular the tagging section. We need to be able to document the community discussion, so the forum or the mail list would be acceptable venues.

Here is the draft: https://wiki.openstreetmap.org/wiki/Import/Catalogue/Sweden_highway_import

The review itself is usually not a problem for this kind of data, but I think someone from the Swedish community should run the process (i.e. not myself). Happy to assist of course.

**Comment by atorger** [comment]
I'm quite sure I've fixed the bugs that caused all of the above failures, and also found a railway crossing handling bug. It did however lead to yet another couple of changes to the endpoint snapping, which by recent history has proven to often cause failures to appear on other files, so no guarantees yet.

I've made test-runs on a handful of municipalities.

Anyway should be ready for another batch run. If these batch runs are very cumbersome to run, I could setup a better pre-testing system here and run more files locally.

**Comment by matthiasfeist** [comment]
@atorger the batch runs are not cumbersome, I essentially just start a amazon EC2 instance and then it chews through the data over night.
I took some time to create a little frontend for the data that is being generated so that it's more usable: http://nvdb-osm-map-data.s3-website.eu-north-1.amazonaws.com/ this always renders the latest data and provides download links to everything. 
It's pretty simple so input is welcome :)

@NKAmapper should we break out this discussion about the "import" plan into a different thread somewhere? I have very little experience with the osm community and it seems that @atorger had some bad experiences and might not be up to drive this... (am I correct?)

**Comment by atorger** [comment]
@NKAmapper nice! I think we can discuss import process under the other open issue "import process"
https://github.com/atorger/nvdb2osm/issues/6
so we can keep this thread for bugs related to the batch runs of all municipalities

I'll put in some more comments there

**Comment by matthiasfeist** [comment]
The batch job has completed and it looks like all kommuner were processed correctly. See here: http://nvdb-osm-map-data.s3-website.eu-north-1.amazonaws.com/
This batch job will from now on run automatically every month on the 1st and download the latest version of the zip from Lastkajen and then run the version on the main branch of your script.

@NKAmapper if you want me to also process the data for Norway in this script and offer them as an easy download, let me know. I think this should be super easy to do, I would just need a list with all the municipalities in Norway and the ids to call your script with...

**Comment by NKAmapper** [comment]
@matthiasfeist: Nice. Thank you for offering for Norway, but I already have a simple process to generate all. Also, nvdb2osm in Norway loads the data itself and runs in a few seconds for each municipality, so it is less of a hassle. The import in Norway is almost completed anyway.

Would it be possible for you to run [this script for Sweden](https://github.com/NKAmapper/nvdb2osm/blob/master/README_SWEDEN.md) for a few more municipalities? Just for illustrating alternative tagging.

**Comment by matthiasfeist** [comment]
@NKAmapper maybe we can break out that discussion into a different thread in your repo? What's the difference between your script and @atorger's? 
About Lastkajen Logistic: It seems there is no way to actually generate an order programatically for the _homogeniserad_ format... 😕

**Comment by NKAmapper** [comment]
@matthiasfeist No problem. There is already a separate item #10.

**Comment by matthiasfeist** [comment]
Hi @atorger and @NKAmapper.
So you know what I am getting at with my pipeline and the automatic download and processing of the files. As I wrote before, I want to have a semi-current visual diff between what's in NVDB and what's in OSM. Here's what I have achieved so far:
<img width="1355" alt="Screenshot 2021-03-14 at 19 55 06" src="https://user-images.githubusercontent.com/1589740/111080601-7e92aa00-84ff-11eb-9deb-ee1327c15f3c.png">

As you can see, there are a couple of layers that show if there are major roads missing, minor roads missing (tracks, service etc), if the highway type is different as well as maxspeed and lane count.

I'm showing this intentionally as hexagons since I don't want to show exactly on the map what might be wrong but rather have a map that shows the general area that the mappers themselves then need to fix manually.

What do you think so far of this? Could this be helpful? My thinking was to have a map where I can see where the most action is needed... And even if I'm not comfortable with adjusting all the geometries and stuff, you could still start with adjusting the lane count for example...

**Comment by atorger** [comment]
That looks really nice!

I don't do spot updates myself, my goal is to get a whole municipality "in synch", so I work through the whole municipality 100-200 way segments at a time until the whole is complete. I think one problem with OSM is that it's spotty quality and you never really know, and by going through the whole municipality I know that it now has a good baseline.

With that working method it's quite easy to find next candidate municipality to work on. If I want to work on some that have lots of missing geometry I can basically make a compare in JOSM with cyclosm.org (quite good overview) and nvdb layer on top. It is cumbersome and takes some time, but "good enough" for me.

However if I had a tool like yours I would use that instead, but I would still work through the whole municipality. I understand though that not all will want to work that way, and for those that want to make smaller updates where it's most needed your tool seems to be very useful visual help. It will also be good to get a view of general quality. Today noone really knows what the status of Sweden is, as it's so impractical to make a comparison, so we only know about the small parts we've zoomed in and looked at. A tool like this would raise awareness, and hopefully inspire more mappers to contribute.

What I think would be valuable, but maybe a bit harder to implement, is to show where the geometry has large offset errors. It's quite common with older mapped roads that are 20 - 40 meters offset and have long sections just as straight lines as the photos at the time of mapping was cloudy.

(It's quite common that there are unnecessary tags on the roads. As NKA already pointed out my script erroneously added lanes=2 on general roads which is not needed (fixed that today). When I map I see "oneway=no" quite often, which also is an unnecessary tag. But conflicting tags is of course most important to consider, unneccessary but correct tags is not too bad)

**Comment by atorger** [comment]
@matthiasfeist FYI, despite all municipalities passing without error last time I have discovered a couple of bugs where snaps the geometry incorrectly, so the output from the last run is not 100% good.

I'm also updating with lots of NKA's comments though with tagging adjustments so we could wait a while still before making a new batch run. I'll post in this thread when I think there is a suitable time to make a new batch run.

**Comment by matthiasfeist** [comment]
To your question about the geometry being 20-40 m off: My script is rather dumb actually 😅 I use the [H3 indexing system](https://h3geo.org/docs/usecases) to put coordinates into buckets. And there is a point in a bucket for Nvdb but not for OSM then my script will complain. It sometimes happens already with geometry that is just 2 meters off because it happens to fall into an adjacent H3 cell. That means that if something is 20m off it will for sure trigger 😆 

But great to hear that you at least think it would be a useful tool. That's at least some indicator for me :)

Don't worry with the updates of course. I'm happy to re-run the pipeline for you at any time!

**Comment by NKAmapper** [comment]
@matthiasfeist Very nice image!
Maxspeed=* and name=* are the top two mistakes, I think.

**Comment by matthiasfeist** [comment]
Yeah good point. I should add _name_. But didn't @atorger say that nvdb does have rather unstable quality of the names?

**Comment by atorger** [comment]
@matthiasfeist yes. Names on city roads are generally good. As NKA has pointed out some cities make lazy naming of cycleways though by copying road names, which I've filed an issue for (the script could auto-clean those). Names in the forestry network is riddled with spelling mistakes and quirks, it's very nice to have them though so I think they are valuable but I do quite some hand fixing of the names. I also think that if some of the misspelled names or naming quirks get in it's still better than unnamed.

In small villages there are often quirks that maybe 1 or 2 driveways get the name of the village and the other are unnamed. Then I generally remove the names.

OSM by tradition uses more of local "unofficial" names and there are some naming differences, so where things are named already it's quite common with conflicts which need to be handled with care.

So we should probably write down a guideline how to manage names, but as said I don't think it's a disaster if these names get in unmodified either.

**Comment by matthiasfeist** [comment]
Ahhh OK I understand. Thanks for the explanation

**Comment by matthiasfeist** [comment]
if someone is interested: I've now hacked together a map that shows the spots where the maxspeed attribute differs between the output of the nvdb2osm script and the current state of open street map: https://nvdb-osm-map-data.s3.eu-north-1.amazonaws.com/map.html#9.58/59.2694/18.0014

I assume that I still have quite some bugs in my script, but I looks promising for now and I have a pipeline where I can in under 15 minutes re-render the whole of sweden to show on the map

**Comment by atorger** [comment]
Nice concept, seems to work really well.

**Comment by NKAmapper** [comment]
Very nice! Some maxspeed updating to be done, it seems.

**Comment by atorger** [comment]
By the way, it seems Trafikverket may be releasing "NVDB-VISQ" a web-based quality visualization tool for NVDB data, to see where in NVDB data there are quality issues like missing maxspeeds etc, maybe something to be inspired by:

https://www.trafikverket.se/contentassets/f6f18881920346a0941846356a1d3a3e/webbinarie-for-externa-utan-anteckningar-2021-01-03---alla.pdf

There's also a new Lastkajen coming soon as already noted by NKA, but it seems it (fortunately) won't mean much change from the current one when it comes to this script.

**Comment by matthiasfeist** [comment]
very interesting. thanks for sharing

**Comment by atorger** [comment]
This issue is getting a bit old and the original problem is resolved, so closing it. But of course, with the new lastkajen we need to do something about splitting of the files, but that is handled in a separate issue.
